power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -273,41 +366,550 @@ inline auto rank_transformers(State const& state) -> RankedTransformerGroups {
+    }
+
+    auto try_calculation_with_regulation(State const& state,
+                                         std::vector<std::vector<RegulatedTransformer>> const& regulator_order,
+                                         CalculationMethod method) const -> ResultType {
+        auto result = calculate_with_retry(state, method);
+
+        bool tap_changed = true;
+        while (tap_changed) {
+            tap_changed = false;
+            UpdateBuffer update_data;
+
+            for (auto const& same_rank_regulators : regulator_order) {
+                for (auto const& regulator : same_rank_regulators) {
+                    if (regulator.transformer.connected_at_tap_side() &&
+                        regulator.transformer.connected_at_control_side(regulator.regulator)) {
+                        tap_changed = tap_changed || control_transformer(regulator, state, result, update_data);
+                    }
+                }
+                if (tap_changed) {
+                    break;
+                }
+            }
+            if (tap_changed) {
+                update_state(update_data);
+                result = calculate_(state, method);
+            }
+        }
+
+        add_tap_regulator_output(state, regulator_order, result);
+
+        return result;
+    }
+
+    auto calculate_with_retry(State const& state, CalculationMethod method) const {
+        try {
+            return calculate_(state, method);
+        } catch (SparseMatrixError const&) {
+            return calculate_(state, CalculationMethod::linear);
+        } catch (IterationDiverge const&) {
+            return calculate_(state, CalculationMethod::linear);
+        }
+    }
+
+    bool control_transformer(RegulatedTransformer const& regulator, State const& state, ResultType const& math_output,
+                             UpdateBuffer& update_data) const {
+        bool tap_changed = false;
+
+        regulator.transformer.apply([&](transformer_c auto const& transformer) {
+            using TransformerType = std::remove_cvref_t<decltype(transformer)>;
+
+            auto const param = regulator.regulator.get().template calc_param<typename ResultType::value_type::sym>();
+
+            auto const& u_node = u_pu_controlled_node<TransformerType>(regulator, state, math_output);
+            auto const& i_node = i_pu_controlled_node<TransformerType>(regulator, state, math_output);
+            auto const u_measured = u_node + param.z_compensation * i_node;
+
+            auto const v_measured = mean_val(cabs(u_measured)); // TODO(mgovers): handle asym correctly
+
+            auto new_tap_pos = [&transformer, &v_measured, &param] {
+                if (v_measured > param.u_set + 0.5 * param.u_band) {
+                    return tap_one_step_up(transformer);
+                }
+                if (v_measured < param.u_set - 0.5 * param.u_band) {
+                    return tap_one_step_down(transformer);
+                }
```

Comment:
This relates to the comment above. And, since the default condition considered in speak is that `tap_max > tap_min`, shouldn't it be:
```suggestion
                if (v_measured > param.u_set + 0.5 * param.u_band) {
                    return tap_one_step_down(transformer);
                }
                if (v_measured < param.u_set - 0.5 * param.u_band) {
                    return tap_one_step_up(transformer);
                }
```

Reply:
also, let's make a `Idx2D .u_set_range` already ;)

Reply:
1. shouldn't be an `Idx2D` (cause it's not an index and not even discrete)
2. we could make it a struct but this directly involves the calculation params and that already is basically that same struct 😬 

what is it you're looking fore here?

Reply:
> what is it you're looking fore here?

First comment


> 1. shouldn't be an `Idx2D` (cause it's not an index and not even discrete)
> 2. we could make it a struct but this directly involves the calculation params and that already is basically that same struct 😬
> 

Minor details. Whatever type that can make a `.u_set_range` to save this:
```C++
param.u_set +/- 0.5 * param.u_band
```

Reply:
done. hope you agree with the new implementation